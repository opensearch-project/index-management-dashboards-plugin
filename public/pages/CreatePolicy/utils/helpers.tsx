/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { EuiText } from "@elastic/eui";

// A helper function to generate a simple string explaining how many elements a user can add to a list.
export const inputLimitText = (currCount = 0, limit = 0, singularKeyword = "", pluralKeyword = "", styleProps = {}) => {
  const difference = limit - currCount;
  const remainingLimit = `You can add up to ${difference} ${limit === 1 ? "" : "more"} ${
    difference === 1 ? singularKeyword : pluralKeyword
  }.`;
  const reachedLimit = `You have reached the limit of ${limit} ${limit === 1 ? singularKeyword : pluralKeyword}.`;
  return (
    <EuiText color={"subdued"} size={"xs"} style={styleProps}>
      {difference > 0 ? remainingLimit : reachedLimit}
    </EuiText>
  );
};
