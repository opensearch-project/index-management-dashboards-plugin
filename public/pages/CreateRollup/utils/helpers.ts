/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { FieldItem } from "../../../../models/interfaces";
import { COMPARISON_OPERATORS } from "./constants";
import moment from "moment";

const timeunits = {
  isMilliseconds: function (stringToCheck: string): boolean {
    return stringToCheck.trim().toLowerCase() == "milliseconds" || stringToCheck.trim().toLowerCase() == "ms";
  },
  isSeconds: function (stringToCheck: string): boolean {
    return stringToCheck.trim().toLowerCase() == "seconds" || stringToCheck.trim().toLowerCase() == "s";
  },
  isMinutes: function (stringToCheck: string): boolean {
    return stringToCheck.trim().toLowerCase() == "minutes" || stringToCheck.trim() == "m";
  },
  isHours: function (stringToCheck: string): boolean {
    return stringToCheck.trim().toLowerCase() == "hours" || stringToCheck.trim().toLowerCase() == "h";
  },
  isDays: function (stringToCheck: string): boolean {
    return stringToCheck.trim().toLowerCase() == "days" || stringToCheck.trim().toLowerCase() == "d";
  },
  isWeeks: function (stringToCheck: string): boolean {
    return stringToCheck.trim().toLowerCase() == "weeks" || stringToCheck.trim().toLowerCase() == "w";
  },
  isMonths: function (stringToCheck: string): boolean {
    return stringToCheck.trim().toLowerCase() == "months" || stringToCheck.trim() == "M";
  },
  isQuarters: function (stringToCheck: string): boolean {
    return stringToCheck.trim().toLowerCase() == "quarters" || stringToCheck.trim().toLowerCase() == "q";
  },
  isYears: function (stringToCheck: string): boolean {
    return stringToCheck.trim().toLowerCase() == "years" || stringToCheck.trim().toLowerCase() == "y";
  },
} as const;

export const buildIntervalScheduleText = (continuous: boolean, interval: number, intervalTimeUnit: string): string => {
  let scheduleText = "";
  if (continuous) {
    scheduleText += "Continuous, every " + interval + " " + parseTimeunit(intervalTimeUnit);
  } else {
    scheduleText += "Not continuous";
  }
  return scheduleText;
};

export const buildCronScheduleText = (continuous: boolean, cronExpression: string): string => {
  let scheduleText = "";
  if (continuous) {
    scheduleText += "Continuous, defined by cron expression: " + cronExpression;
  } else {
    scheduleText += "Not continuous";
  }
  return scheduleText;
};

export const parseTimeunit = (timeunit: string): string => {
  if (timeunits.isMilliseconds(timeunit)) return "millisecond(s)";
  else if (timeunits.isSeconds(timeunit)) return "second(s)";
  else if (timeunits.isMinutes(timeunit)) return "minute(s)";
  else if (timeunits.isHours(timeunit)) return "hour(s)";
  else if (timeunits.isDays(timeunit)) return "day(s)";
  else if (timeunits.isWeeks(timeunit)) return "week";
  else if (timeunits.isMonths(timeunit)) return "month";
  else if (timeunits.isQuarters(timeunit)) return "quarter";
  else if (timeunits.isYears(timeunit)) return "year";

  return timeunit;
};

export const delayTimeUnitToMS = (delay: number, timeunit: string): number => {
  if (timeunits.isSeconds(timeunit)) {
    return moment.duration(delay, "seconds").asMilliseconds();
  } else if (timeunits.isMinutes(timeunit)) {
    return moment.duration(delay, "minutes").asMilliseconds();
  } else if (timeunits.isHours(timeunit)) {
    return moment.duration(delay, "hours").asMilliseconds();
  } else if (timeunits.isDays(timeunit)) {
    return moment.duration(delay, "days").asMilliseconds();
  } else {
    return delay;
  }
};

export const msToDelayTimeUnit = (delay: number, timeunit: string): number => {
  if (timeunits.isSeconds(timeunit)) {
    return moment.duration(delay, "milliseconds").asSeconds();
  } else if (timeunits.isMinutes(timeunit)) {
    return moment.duration(delay, "milliseconds").asMinutes();
  } else if (timeunits.isHours(timeunit)) {
    return moment.duration(delay, "milliseconds").asHours();
  } else if (timeunits.isDays(timeunit)) {
    return moment.duration(delay, "milliseconds").asDays();
  } else {
    return delay;
  }
};

//Returns true if field type is numeric
export const isNumericMapping = (fieldType: string | undefined): boolean => {
  return (
    fieldType == "long" ||
    fieldType == "integer" ||
    fieldType == "short" ||
    fieldType == "byte" ||
    fieldType == "double" ||
    fieldType == "float" ||
    fieldType == "half_float" ||
    fieldType == "scaled_float"
  );
};

export const compareFieldItem = (itemA: FieldItem, itemB: FieldItem): boolean => {
  return itemB.label == itemA.label && itemA.type == itemB.type;
};

export const parseFieldOptions = (prefix: string, mappings: any): FieldItem[] => {
  let fieldsOption: FieldItem[] = [];
  for (let field in mappings) {
    if (mappings.hasOwnProperty(field)) {
      if (mappings[field].type != "object" && mappings[field].type != null && mappings[field].type != "nested")
        fieldsOption.push({ label: prefix + field, type: mappings[field].type });
      if (mappings[field].fields != null)
        fieldsOption = fieldsOption.concat(parseFieldOptions(prefix + field + ".", mappings[field].fields));
      if (mappings[field].properties != null)
        fieldsOption = fieldsOption.concat(parseFieldOptions(prefix + field + ".", mappings[field].properties));
    }
  }
  return fieldsOption;
};

export const getOperators = (fieldType) =>
  COMPARISON_OPERATORS.reduce(
    (acc, currentOperator) =>
      currentOperator.dataTypes.includes(fieldType) ? [...acc, { text: currentOperator.text, value: currentOperator.value }] : acc,
    []
  );
