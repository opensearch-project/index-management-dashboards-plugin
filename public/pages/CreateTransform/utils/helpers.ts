/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { FieldItem, TRANSFORM_AGG_TYPE, TransformGroupItem } from "../../../../models/interfaces";
import { IntervalType } from "../../../utils/constants";

export const parseTimeunit = (timeunit: string): string => {
  if (timeunit === "ms" || timeunit === "Milliseconds") return "millisecond(s)";
  else if (timeunit === "SECONDS" || timeunit === "s" || timeunit === "Seconds") return "second(s)";
  else if (timeunit === "MINUTES" || timeunit === "m" || timeunit === "Minutes") return "minute(s)";
  else if (timeunit === "HOURS" || timeunit === "h" || timeunit === "Hours") return "hour(s)";
  else if (timeunit === "DAYS" || timeunit === "d" || timeunit === "Days") return "day(s)";
  else if (timeunit === "w") return "week";
  else if (timeunit === "M") return "month";
  else if (timeunit === "q") return "quarter";
  else if (timeunit === "y") return "year";

  return timeunit;
};

// Returns true if field type is numeric
export const isNumericMapping = (fieldType: string | undefined): boolean => {
  return (
    fieldType === "long" ||
    fieldType === "integer" ||
    fieldType === "short" ||
    fieldType === "byte" ||
    fieldType === "double" ||
    fieldType === "float" ||
    fieldType === "half_float" ||
    fieldType === "scaled_float"
  );
};

export const compareFieldItem = (itemA: FieldItem, itemB: FieldItem): boolean => {
  return itemB.label === itemA.label && itemA.type === itemB.type;
};

export const parseFieldOptions = (prefix: string, mappings: any): FieldItem[] => {
  let fieldsOption: FieldItem[] = [];
  for (const field in mappings) {
    if (mappings.hasOwnProperty(field)) {
      if (mappings[field].type !== "object" && mappings[field].type != null && mappings[field].type !== "nested")
        fieldsOption.push({ label: prefix + field, type: mappings[field].type, path: mappings[field].path });
      if (mappings[field].fields != null)
        fieldsOption = fieldsOption.concat(parseFieldOptions(prefix + field + ".", mappings[field].fields));
      if (mappings[field].properties != null)
        fieldsOption = fieldsOption.concat(parseFieldOptions(prefix + field + ".", mappings[field].properties));
    }
  }
  return fieldsOption;
};

export const createdTransformToastMessage = (transformId: string): string => {
  return `Transform job "${transformId}" successfully created.`;
};

/**
 * Searches the stringToSearch for all occurrences of the transformId,
 * and wraps each occurrence with opening and closing quotation marks.
 * @param transformId The string that serves as the transform's job name.
 * @param stringToSearch The string to search.
 * @return The stringToSearch but with any occurrences of transformId wrapped in quotation marks.
 */
export const wrapQuotesAroundTransformId = (transformId: string, stringToSearch: string): string => {
  const regex = new RegExp(transformId, "g");
  const idWrappedWithQuotes = `"${transformId}"`;
  return stringToSearch.replace(regex, idWrappedWithQuotes);
};

export const isGroupBy = (type: string): boolean => {
  return type === TRANSFORM_AGG_TYPE.histogram || type === TRANSFORM_AGG_TYPE.terms || type === TRANSFORM_AGG_TYPE.date_histogram;
};

export const getDateHistogramGroupItem = (
  name: string,
  targetFieldName: string,
  interval: number,
  timeunit: string,
  intervalType: IntervalType
): TransformGroupItem => {
  const dateHistogramInterval = `${interval}${timeunit}`;
  if (intervalType === IntervalType.CALENDAR)
    return {
      date_histogram: {
        source_field: name,
        target_field: targetFieldName,
        calendar_interval: dateHistogramInterval,
      },
    };
  else
    return {
      date_histogram: {
        source_field: name,
        target_field: targetFieldName,
        fixed_interval: dateHistogramInterval,
      },
    };
};
