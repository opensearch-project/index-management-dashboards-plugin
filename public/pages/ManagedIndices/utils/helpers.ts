/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import queryString from "query-string";
import { DEFAULT_QUERY_PARAMS } from "./constants";
import { ManagedIndicesQueryParams } from "../models/interfaces";

export function getURLQueryParams(location: { search: string }): ManagedIndicesQueryParams {
  const { from, size, search, sortField, sortDirection, showDataStreams } = queryString.parse(location.search);

  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  return <ManagedIndicesQueryParams>{
    // @ts-ignore
    from: isNaN(parseInt(from, 10)) ? DEFAULT_QUERY_PARAMS.from : parseInt(from, 10),
    // @ts-ignore
    size: isNaN(parseInt(size, 10)) ? DEFAULT_QUERY_PARAMS.size : parseInt(size, 10),
    search: typeof search !== "string" ? DEFAULT_QUERY_PARAMS.search : search,
    sortField: typeof sortField !== "string" ? DEFAULT_QUERY_PARAMS.sortField : sortField,
    sortDirection: typeof sortDirection !== "string" ? DEFAULT_QUERY_PARAMS.sortDirection : sortDirection,
    showDataStreams: showDataStreams === undefined ? DEFAULT_QUERY_PARAMS.showDataStreams : "true" === showDataStreams,
  };
}
