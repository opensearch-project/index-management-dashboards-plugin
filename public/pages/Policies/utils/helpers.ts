/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import queryString from "query-string";
// @ts-ignore
import moment from "moment";
import { DEFAULT_QUERY_PARAMS } from "./constants";
import { PoliciesQueryParams } from "../models/interfaces";

export function getURLQueryParams(location: { search: string }): PoliciesQueryParams {
  const { from, size, search, sortField, sortDirection } = queryString.parse(location.search);

  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  return <PoliciesQueryParams>{
    // @ts-ignore
    from: isNaN(parseInt(from, 10)) ? DEFAULT_QUERY_PARAMS.from : parseInt(from, 10),
    // @ts-ignore
    size: isNaN(parseInt(size, 10)) ? DEFAULT_QUERY_PARAMS.size : parseInt(size, 10),
    search: typeof search !== "string" ? DEFAULT_QUERY_PARAMS.search : search,
    sortField: typeof sortField !== "string" ? DEFAULT_QUERY_PARAMS.sortField : sortField,
    sortDirection: typeof sortDirection !== "string" ? DEFAULT_QUERY_PARAMS.sortDirection : sortDirection,
  };
}

export const renderTime = (time: number): string => {
  const momentTime = moment(time).local();
  if (time && momentTime.isValid()) return momentTime.format("MM/DD/YY h:mm a");
  return "-";
};
