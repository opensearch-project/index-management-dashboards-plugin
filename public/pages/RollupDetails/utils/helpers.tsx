/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { EuiFlexGroup, EuiFlexItem, EuiIcon, EuiText } from "@elastic/eui";
import { RollupMetadata } from "../../../../models/interfaces";

export const renderStatus = (metadata: RollupMetadata | undefined): JSX.Element => {
  if (metadata == null || metadata === undefined || metadata.rollup_metadata == null) return <dd>-</dd>;
  let icon;
  let iconColor;
  let textColor: "default" | "subdued" | "secondary" | "ghost" | "accent" | "warning" | "danger" | undefined;
  let text;
  switch (metadata.rollup_metadata.status) {
    case "failed":
      icon = "alert";
      iconColor = "danger";
      textColor = "danger";
      text = "Failed: " + metadata.rollup_metadata.failure_reason;
      break;
    case "finished":
      icon = "check";
      iconColor = "success";
      textColor = "secondary";
      text = "Complete";
      break;
    case "init":
      return (
        <EuiFlexGroup gutterSize="xs">
          <EuiFlexItem grow={false}>
            <EuiIcon size="s" type="clock" color="primary" />
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <EuiText size="xs" style={{ color: "#006BB4" }}>
              Initializing
            </EuiText>
          </EuiFlexItem>
        </EuiFlexGroup>
      );
    case "started":
      icon = "play";
      iconColor = "success";
      textColor = "secondary";
      text = "Started";
      break;
    case "stopped":
      icon = "stop";
      iconColor = "subdued";
      textColor = "subdued";
      text = "Stopped";
      break;
    default:
      return <dd>-</dd>;
  }

  return (
    <EuiFlexGroup gutterSize="xs">
      <EuiFlexItem grow={false}>
        <EuiIcon size="s" type={icon} color={iconColor} />
      </EuiFlexItem>
      <EuiFlexItem grow={false}>
        <EuiText size="xs" color={textColor}>
          {text}
        </EuiText>
      </EuiFlexItem>
    </EuiFlexGroup>
  );
};
