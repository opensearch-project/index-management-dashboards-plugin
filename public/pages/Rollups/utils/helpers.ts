/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import queryString from "query-string";
// @ts-ignore
import moment from "moment";
import { DEFAULT_QUERY_PARAMS } from "./constants";
import { RollupQueryParams } from "../models/interfaces";

export function getURLQueryParams(location: { search: string }): RollupQueryParams {
  const { from, size, search, sortField, sortDirection } = queryString.parse(location.search);

  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  return <RollupQueryParams>{
    // @ts-ignores
    from: isNaN(parseInt(from, 10)) ? DEFAULT_QUERY_PARAMS.from : parseInt(from, 10),
    // @ts-ignore
    size: isNaN(parseInt(size, 10)) ? DEFAULT_QUERY_PARAMS.size : parseInt(size, 10),
    search: typeof search !== "string" ? DEFAULT_QUERY_PARAMS.search : search,
    sortField: typeof sortField !== "string" ? DEFAULT_QUERY_PARAMS.sortField : sortField,
    sortDirection: typeof sortDirection !== "string" ? DEFAULT_QUERY_PARAMS.sortDirection : sortDirection,
  };
}

export const renderTime = (time: number): string => {
  const momentTime = moment(time).local();
  if (time && momentTime.isValid()) return momentTime.format("MM/DD/YY h:mm a");
  return "-";
};

export const renderEnabled = (isEnabled: boolean): string => {
  return isEnabled ? "Enabled" : "Disabled";
};

export const renderContinuous = (continuous: boolean): string => {
  return continuous ? "Yes" : "No";
};
