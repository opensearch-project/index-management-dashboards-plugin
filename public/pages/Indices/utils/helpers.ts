/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import queryString from "query-string";
import { CoreStart } from "opensearch-dashboards/public";
import { DEFAULT_QUERY_PARAMS } from "./constants";
import { IndicesQueryParams } from "../models/interfaces";
import { IndexItem } from "../../../../models/interfaces";
import { ServerResponse } from "../../../../server/models/types";
import { CommonService } from "../../../services";
import { CatIndex } from "../../../../server/models/interfaces";
import { jobSchedulerInstance } from "../../../context/JobSchedulerContext";
import { OpenJobMetaData, RecoveryJobMetaData } from "../../../models/interfaces";
import { ListenType } from "../../../lib/JobScheduler";
import { getClusterInfo } from "../../../utils/helpers";

export function getURLQueryParams(location: { search: string }): IndicesQueryParams {
  const { from, size, search, sortField, sortDirection, showDataStreams } = queryString.parse(location.search);
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  return <IndicesQueryParams>{
    // @ts-ignore
    from: isNaN(parseInt(from, 10)) ? DEFAULT_QUERY_PARAMS.from : parseInt(from, 10),
    // @ts-ignore
    size: isNaN(parseInt(size, 10)) ? DEFAULT_QUERY_PARAMS.size : parseInt(size, 10),
    search: typeof search !== "string" ? DEFAULT_QUERY_PARAMS.search : search,
    sortField: typeof sortField !== "string" ? "index" : sortField,
    sortDirection: typeof sortDirection !== "string" ? DEFAULT_QUERY_PARAMS.sortDirection : sortDirection,
    showDataStreams: showDataStreams === undefined ? DEFAULT_QUERY_PARAMS.showDataStreams : showDataStreams === "true",
  };
}

export async function openIndices(props: {
  indices: string[];
  commonService: CommonService;
  coreServices: CoreStart;
  jobConfig?: Partial<OpenJobMetaData>;
}) {
  const indexPayload = props.indices.join(",");
  const result = await props.commonService.apiCaller<{
    task: string;
  }>({
    endpoint: "transport.request",
    data: {
      method: "POST",
      path: `/${indexPayload}/_open?wait_for_completion=false`,
    },
  });

  if (result && result.ok) {
    const toast = `Successfully started opening ${indexPayload}.`;
    const toastInstance = props.coreServices.notifications.toasts.addSuccess(toast, {
      toastLifeTimeMs: 1000 * 60 * 60 * 24 * 5,
    });
    const clusterInfo = await getClusterInfo({
      commonService: props.commonService,
    });
    await jobSchedulerInstance.addJob({
      type: ListenType.OPEN,
      extras: {
        clusterInfo,
        toastId: toastInstance.id,
        taskId: result.response?.task,
        indexes: props.indices,
      },
      interval: 3000,
      ...props.jobConfig,
    } as OpenJobMetaData);
  } else {
    props.coreServices.notifications.toasts.addDanger(result.error);
  }

  return result;
}

export async function getIndexSettings(props: {
  indexName: string;
  flat: boolean;
  commonService: CommonService;
  coreServices: CoreStart;
}): Promise<Record<string, IndexItem>> {
  const result: ServerResponse<Record<string, IndexItem>> = await props.commonService.apiCaller({
    endpoint: "indices.getSettings",
    method: "GET",
    data: {
      index: props.indexName,
      flat_settings: props.flat,
    },
  });
  if (result && result.ok) {
    return result.response;
  } else {
    const errorMessage = `There is a problem getting index setting for ${props.indexName}, please check with Admin`;
    props.coreServices.notifications.toasts.addDanger(result?.error || errorMessage);
    throw new Error(result?.error || errorMessage);
  }
}

export async function getSingleIndice(props: {
  indexName: string;
  commonService: CommonService;
  coreServices: CoreStart;
}): Promise<CatIndex> {
  const result = await props.commonService.apiCaller({
    endpoint: "cat.indices",
    method: "GET",
    data: {
      format: "json",
      index: [`${props.indexName || ""}`],
    },
  });

  if (result && result.ok && result.response.length === 1) {
    return result.response[0];
  } else {
    const errorMessage = `There is a problem getting index for ${props.indexName}, please check with Admin`;
    props.coreServices.notifications.toasts.addDanger(result?.error || errorMessage);
    throw new Error(result?.error || errorMessage);
  }
}

export async function setIndexSettings(props: {
  indexName: string;
  flat: boolean;
  settings: {};
  commonService: CommonService;
  coreServices: CoreStart;
}) {
  const result = await props.commonService.apiCaller({
    endpoint: "indices.putSettings",
    method: "PUT",
    data: {
      index: props.indexName,
      flat_settings: props.flat,
      body: {
        settings: {
          ...props.settings,
        },
      },
    },
  });
  if (result && result.ok) {
    props.coreServices.notifications.toasts.addSuccess(`Successfully update index setting for ${props.indexName}`);
  } else {
    const errorMessage = `There is a problem set index setting for ${props.indexName}, please check with Admin`;
    props.coreServices.notifications.toasts.addDanger(result?.error || errorMessage);
    throw new Error(result?.error || errorMessage);
  }
}

export async function getAlias(props: { aliasName?: string; commonService: CommonService }) {
  return await props.commonService.apiCaller<Array<{ alias: string }>>({
    endpoint: "cat.aliases",
    method: "GET",
    data: {
      format: "json",
      name: `*${props.aliasName || ""}*`,
      s: "alias:desc",
    },
  });
}

export async function splitIndex(props: {
  sourceIndex: string;
  targetIndex: string;
  settingsPayload: Required<IndexItem>["settings"];
  commonService: CommonService;
  coreServices: CoreStart;
}) {
  const { aliases, ...settings } = props.settingsPayload;
  const result = await props.commonService.apiCaller<{
    task: string;
  }>({
    endpoint: "transport.request",
    data: {
      path: `/${props.sourceIndex}/_split/${props.targetIndex}?wait_for_completion=false`,
      method: "PUT",
      body: {
        settings: {
          ...settings,
        },
        aliases,
      },
    },
  });

  if (result && result.ok) {
    const toastInstance = props.coreServices.notifications.toasts.addSuccess(
      `Successfully started splitting ${props.sourceIndex}. The split index will be named ${props.targetIndex}`,
      {
        toastLifeTimeMs: 1000 * 60 * 60 * 24 * 5,
      }
    );
    const clusterInfo = await getClusterInfo({
      commonService: props.commonService,
    });
    await jobSchedulerInstance.addJob({
      interval: 30000,
      extras: {
        clusterInfo,
        toastId: toastInstance.id,
        sourceIndex: props.sourceIndex,
        destIndex: props.targetIndex,
        taskId: result.response?.task,
      },
      type: ListenType.SPLIT,
    } as RecoveryJobMetaData);
    return result;
  } else {
    const errorMessage = `There was a problem submit split index request, please check with admin`;
    props.coreServices.notifications.toasts.addDanger(result?.error || errorMessage);
    throw new Error(result?.error || errorMessage);
  }
}

/*
 ** When splitting an index, the valid value of number_of_shards in the target index depends on the source index’s primary shards count:
 ** (1) If the source index’s primary shards count is equal to 1,
 **     then the valid value is an arbitrary number between 1 and 1024(includes 1 and 1024).
 ** (2) If the source index’s primary shards count is greater than 1,
 **     the valid value is obtained by continuously multiplying the source index’s primary shards count by 2
 **     as long as it's not larger than 1024.
 **     For example, if the source index has 3 primary shards, then all the valid value are 3,6,12,24,…,768.
 */
export function getSplitShardOptions(sourceShards: number) {
  const MAX_SHARDS_NUMBER = 1024;
  const shardsSelectOptions = [];
  if (sourceShards === 1) {
    for (let i = 2; i <= MAX_SHARDS_NUMBER; i++) {
      shardsSelectOptions.push({
        label: i.toString(),
      });
    }
  } else {
    const SHARDS_HARD_LIMIT = MAX_SHARDS_NUMBER / 2;
    let shardsLimit = sourceShards;
    while (shardsLimit <= SHARDS_HARD_LIMIT) {
      shardsLimit = shardsLimit * 2;
      shardsSelectOptions.push({
        label: shardsLimit.toString(),
      });
    }
  }

  return shardsSelectOptions;
}
