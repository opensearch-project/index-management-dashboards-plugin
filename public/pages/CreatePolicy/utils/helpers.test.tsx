/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { render, screen } from "@testing-library/react";
import { inputLimitText } from "./helpers";

function renderInputLimitText(currCount?, limit?, singularKeyword?, pluralKeyword?, styleProps?) {
  return { ...render(inputLimitText(currCount, limit, singularKeyword, pluralKeyword, styleProps)) };
}

describe("inputLimitText", () => {
  it("renders the component with 0 inputs", () => {
    const expected = `You can add up to 10 more aliases.`;
    const { container } = renderInputLimitText(0, 10, "alias", "aliases");
    expect(screen.getByText(expected)).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders the component with 1 input remaining", () => {
    const expected = "You can add up to 1 more alias.";
    const { container } = renderInputLimitText(9, 10, "alias", "aliases");
    expect(screen.getByText(expected)).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders the component with 0 inputs remaining", () => {
    const expected = "You have reached the limit of 10 aliases.";
    const { container } = renderInputLimitText(10, 10, "alias", "aliases");
    expect(screen.getByText(expected)).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });
});
