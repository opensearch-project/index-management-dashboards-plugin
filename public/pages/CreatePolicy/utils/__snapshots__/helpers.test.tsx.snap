// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inputLimitText renders the component with 0 inputs 1`] = `
<div
  class="euiText euiText--extraSmall"
>
  <div
    class="euiTextColor euiTextColor--subdued"
  >
    You can add up to 10 more aliases.
  </div>
</div>
`;

exports[`inputLimitText renders the component with 0 inputs remaining 1`] = `
<div
  class="euiText euiText--extraSmall"
>
  <div
    class="euiTextColor euiTextColor--subdued"
  >
    You have reached the limit of 10 aliases.
  </div>
</div>
`;

exports[`inputLimitText renders the component with 1 input remaining 1`] = `
<div
  class="euiText euiText--extraSmall"
>
  <div
    class="euiTextColor euiTextColor--subdued"
  >
    You can add up to 1 more alias.
  </div>
</div>
`;
